[Model Loading] Speedup model loading with distributed loading

chestnut-Q opened the pull request:
Hello! The current method for model loading is quite fixed, regardless of the tensor parallel size. It involves each rank in a tp group reading the full weight file, and then discarding the excess weight tensors if only a portion of the parameters is needed for that rank. When `--tensor-parallel-size` is greater than 1, most parameters require only 1/tp_size of the parameters, leading to significant additional weights IO.

Observing that the disk IO speed is slow (particularly for bin files), and the transfer rate between multiple GPUs is fast, we can adopt a distributed loading approach. This means each worker loads only 1/tp_size of the weight file (by file division, or for SafeTensors type, it can be by tensor division). Then the parameters needed by the workers are transferred to each other using `torch.distributed.scatter` or `torch.distributed.broadcast`. This approach can reduce disk IO to 1/tp_size.

I have implemented the example distributed loading code in `llama.py` and `baichuan.py`. I believe other models (if needed) can easily implement similar logic. To ensure compatibility with previous codes, the args introduced in this PR are optional. Therefore, if you do not wish to use distributed loading, the original code does not require any modifications.

When --tensor-parallel-size >= 4, the distributed loading method can significantly accelerate loading times, typically by 40% or more. Here are the experiment results on my machine (8*A100) for Llama-2-70b and Baichuan2-13B.

|             | Llama-2-70b (TP8) | Baichuan2-13B (TP4) |
| ----------- | ----------------- | ------------------- |
| Vanilla     | 249.5s            | 45.3s               |
| Distributed | 141.5s            | 25.2s               |
| Speedup     | 43.3%             | 44.4%               |

chestnut-Q committed:
Distributed weight loading

ywang96 commented on vllm/model_executor/layers/activation.py:
```diff
@@ -127,11 +128,28 @@ def __init__(
         if self.input_is_parallel:
             tp_rank = get_tensor_model_parallel_rank()
             shard_size = param_data.shape[0]
+            if weight_owner is not None:
+                tp_size = get_tensor_model_parallel_world_size()
+                tp_group = get_tensor_model_parallel_group()
+                if tp_rank == torch.distributed.get_group_rank(
+                        tp_group, weight_owner):
+                    loaded_weight_list = list(
+                        torch.split(loaded_weight, shard_size, dim=0))
+                    assert len(loaded_weight_list) == tp_size
+                    assert param_data.shape == loaded_weight_list[0].shape
+                else:
+                    loaded_weight_list = None
+                torch.distributed.scatter(param_data, loaded_weight_list,
+                                          weight_owner, tp_group)
```
It seems to me that this code pattern has been repeated a lot throughout this PR - why not making a separate helper/util function for this or am I missing something here?

chestnut-Q replied:
Fixed

ywang96 commented on vllm/model_executor/models/baichuan.py:
```diff
@@ -351,8 +352,42 @@ def load_weights(self,
+                        cache_dir, load_format, revision)
+                for name in weight_names:
+                    weight_owner = weights_map[name]
+                    loaded_weight = local_state_dict[name].to("cuda") \
```
Why `"cuda"` in particular?

chestnut-Q replied:
Because the loaded_weight is on the cpu after loading, but the params are on the gpu, and the transfer rate of tensors on the gpu is faster.

ywang96 replied:
To clarify - I'm asking why we need to hardcode "cuda" here since vLLM is supported by multiple hardware types. If this would only work on CUDA devices then we might need to do a check at the engine initialization.

chestnut-Q replied:
Got it. I think this approach only works on CUDA devices right now, because the speed of transferring data across multiple ranks might not be fast enough on other devices to be worth it. I've added some checks for this at the engine initialization.

ywang96 commented on vllm/worker/model_runner.py:
```diff
@@ -91,6 +91,7 @@ def __init__(
             self.model_config.dtype if model_config is not None else None)
 
     def load_model(self) -> None:
+        load_model_start = time.time()
```
```suggestion
        load_model_start = time.monotonic()
```

chestnut-Q replied:
Adopted

ywang96 commented on vllm/worker/model_runner.py:
```diff
@@ -102,7 +103,8 @@ def load_model(self) -> None:
         logger.info(f"Loading model weights took "
-                    f"{self.model_memory_usage / float(2**30):.4f} GB")
+                    f"{self.model_memory_usage / float(2**30):.4f} GB "
+                    f"and cost {time.time() - load_model_start:.1f}s")
```
```suggestion
                    f"and {time.monotonic()- load_model_start:.1f}s.")
```

chestnut-Q replied:
Adopted

ywang96 commented on vllm/model_executor/models/baichuan.py:
```diff
@@ -385,7 +423,8 @@ def load_weights(self,
                                         default_weight_loader)
-                weight_loader(param, loaded_weight)
+                weight_loader(param, loaded_weight, weight_owner=weight_owner)
+        torch.cuda.empty_cache()
```
Why do we need to manually call the cache release here in each model file (instead of as a part of the util)?

chestnut-Q replied:
Fixed

ywang96 commented on vllm/model_executor/models/baichuan.py:
```diff
@@ -351,8 +352,42 @@ def load_weights(self,
+                weight_names.append(name)
+        weight_names = sorted(weight_names)
+
+        def weights_iterator(
+            use_distributed_loading: bool
+        ) -> Iterator[Tuple[str, Optional[torch.Tensor], Optional[int]]]:
+            if use_distributed_loading:
+                local_state_dict, weights_map = \
+                    hf_model_weights_distributed_loader(
+                        model_name_or_path, weight_names,
+                        cache_dir, load_format, revision)
+                for name in weight_names:
+                    weight_owner = weights_map[name]
+                    loaded_weight = local_state_dict[name].to("cuda") \
+                        if weight_owner == get_tensor_model_parallel_rank() \
+                        else None
+                    yield name, loaded_weight, weight_owner
+            else:
+                for name, loaded_weight in hf_model_weights_iterator(
+                        model_name_or_path, cache_dir, load_format, revision):
+                    yield name, loaded_weight, None
```
Why does this need to be defined within the model file instead of as a separate helper?

chestnut-Q replied:
Fixed

ywang96 reviewed (changes requested):
Hello! Thanks for the PR and I left some comments & questions. This is indeed a neat feature but please consider the use of utils & helper modules. 

IMO ideally we would like to reduce the need of changing each model file as much as possible to make the codebase easier to maintain.

chestnut-Q commented:
@ywang96 Thank you for your suggestion! I have moved the duplicate codes to utils as per your advice, which resulted in minimal changes to each model file and made them clearer. Additionally, I have added the cli arg `--use-distributed-loading` and expanded support for other models that have a parameter structure similar to llama. If there are any more questions or suggestions, welcome to comment :)

chestnut-Q committed:
Run yapf and ruff

chestnut-Q committed:
Run yapf and ruff

BenNR commented:
Hi guys this is a truly valuable feature is this still moving forward into an official vllm release? Not trying to be pushy great work and cool concept though!

sdake commented:
cc/ @sdake

youkaichao commented:
close as this becomes stale.

and please see https://github.com/vllm-project/vllm/pull/6127#issuecomment-2221614995 

we recommend using safetensors format, and then we don't need this optimization.
